### ISSUE TITLE
Investigate the use of a chatbot to form a knowledgebase

### ISSUE THREAD

**Issue by stichbury** [kedro-org/kedro#2026]
We need a good way for Kedro users to get answers to their questions. Right now, they could search the discord linen archive or Slack for previous discussions but the UX isn't great. Or they can look at our written FAQ (which they do -- it's a popular page) but it doesn't deliver what they expect (it doesn't go into specific answers to specific questions).  They probably end up on Google looking for answers and maybe StackOverflow.

It would be great to run a NLP chatbot that has been trained on our archives and documentation and can have a stab at an answer or link users to the right location to start their research. This is the holy grail of all documentation though, and does rely on a decent knowledgebase to train it, which we probably don't have (at least, we have content, but it's not clear whether it is suitable).

I think we need to first investigate the state of this kind of solution and then look at whether we can apply it to a Slackbot for Kedro.

This is early days, but here's a few links:

* https://ai.google.com/research/NaturalQuestions
* https://towardsdatascience.com/building-a-conversational-chatbot-for-slack-using-rasa-and-python-part-1-bca5cc75d32f (https://rasa.com/)
* https://github.com/juliakreutzer/neural-slack-bot

This issue is to seed some discussion and potentially earmark some time for research at a hackathon or similar spike.

**Comment by stichbury** [comment]
Recently, we've seen the [ChatGPT project](https://chat.openai.com/chat) put out a beta of what I was mumbling about above. Thanks OpenAI 👯 

While it's not yet ready, there are some ways we can prepare for a future where Kedro users turn to ChatGPT to answer their queries. A rough list:

- [ ] We need to ensure that our corpus of Q&A is indexed. This includes linen archives and any other content existing on Slack, Discourse etc.
- [ ] We create some complete and useful answers that rank well on queries such as "What is Kedro?" (and, more generally, "How do I ensure my data science code is reusable" etc. This ensures Kedro gets into the text returned)

**Comment by yetudada** [comment]
Looks like this will be possible, Dagster built something like this: https://dagster.io/blog/chatgpt-langchain

This ticket should also probably link to #1649

**Comment by stichbury** [comment]
@astrojuanlu Also this one up for discussion

**Comment by astrojuanlu** [comment]
Somebody pointed me into this direction: https://langchain.readthedocs.io/en/latest/use_cases/question_answering.html

**Comment by datajoely** [comment]
Relevant tool
https://www.mendable.ai/?s=03

**Comment by astrojuanlu** [comment]
More: https://docsbot.ai/

**Comment by noklam** [comment]
It's crazy how quick this space evolve. It's quite feasible to build one with langchain, you can also limited the context that it reads `docs.kedro.org` and generate answer with relevant link only in the docs (so it's not making random stuff up).

**Comment by astrojuanlu** [comment]
"Don't replace your user community with an LLM-based chatbot" https://thisisimportant.net/posts/user-community-llm-chatbots/

**Comment by astrojuanlu** [comment]
Beware of pushback from the tech community https://github.com/mdn/yari/issues/9208

**Comment by datajoely** [comment]
This is interesting
https://docs.danswer.dev/introduction
https://github.com/danswer-ai/danswer

**Comment by stichbury** [comment]
> "Don't replace your user community with an LLM-based chatbot" https://thisisimportant.net/posts/user-community-llm-chatbots/

I don't think it was ever a binary choice of either/or, was it? If the community want a knowledgebase, let's give it to them alongside current options...we're not proposing to remove anything.

**Comment by stichbury** [comment]
@noklam and I worked on this as part of the Quantazio Hack. It would be good to continue the work as part of the ongoing docs effort.

**Comment by astrojuanlu** [comment]
If we ever get to this, probably we'd use some form of Retrieval Augmented Generation (RAG), see https://github.com/imartinez/privategpt

**Comment by astrojuanlu** [comment]
Or just https://slack.com/intl/en-gb/help/articles/202026038-An-introduction-to-Slackbot#add-customized-automatic-responses

**Comment by astrojuanlu** [comment]
also xref https://github.com/kedro-org/kedro-plugins/pull/434

**Comment by astrojuanlu** [comment]
Related: publishing a custom GPT on Kedro, MLOps? https://help.openai.com/en/articles/8798878-building-and-publishing-a-gpt

**Comment by astrojuanlu** [comment]
This is a worthy solution to the problem of 

> We need a good way for Kedro users to get answers to their questions.

However, there are plenty of other things that can be done before creating our own chatbot:

- #3741 
- #4482
- #4257 

And I could name a few more.

This is not in our short term priorities, so I'm closing for now.
